### DISCUSSION TITLE
LibraryImportAttribute emits DllImport in generated code

### DISCUSSION THREAD

**Discussion by PhilPJL** [dotnet/runtime#77086] [Q&A] [answered]
I've been experimenting changing some p-invoke DllImport statements to use LibraryImport.  For example a simple one:

```
    [DllImport(Constants.DLLFileName, EntryPoint = "H5Pclose",
        CallingConvention = CallingConvention.Cdecl),
    SuppressUnmanagedCodeSecurity, SecuritySafeCritical]
    public static extern herr_t close(hid_t plist);
```

changes to (I know these SuppressUnmanagedCodeSecurity, SecuritySafeCritical are null-ops in .NET core)

```
    [LibraryImport(Constants.DLLFileName, EntryPoint = "H5Pclose"),
    SuppressUnmanagedCodeSecurity, SecuritySafeCritical]
    [UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvCdecl) })]
    public static partial herr_t close(hid_t plist);
```

I'm surprised to see that the generated code still uses DllImport

```
    internal static unsafe partial class H5P
    {
        [System.Runtime.InteropServices.DllImportAttribute("hdf5", EntryPoint = "H5Pclose", ExactSpelling = true)]
        public static extern partial int close(long plist);
    }

```

Is this correct/expected?

**Comment by alexrp** [comment, marked as answer]
Yes. The goal isn't to avoid `DllImport`, it's to avoid runtime marshaling of parameters and return values. The generated `DllImport`-decorated method will only use blittable types, while the wrapper does marshaling through source generation before/after calling it.

**Reply by PhilPJL** [reply]
Ok, I see. Thanks.
